### DISCUSSION TITLE
Dynamic Configuration Service for help

### DISCUSSION THREAD

**Discussion by Jiaweij** [apache/skywalking#12577] [Q&A] [answered]
Dear all
         Recently i  get a job for testing  Dynamic  Configuration Server (gprc) , i read the  guide document (https://skywalking.apache.org/docs/main/v10.0.1/en/setup/backend/dynamic-config-service/) , but  i still has some problem 
         1. i have already write a  grpc server which listen on 5005 port ,  but i did not see where to difine the method name and param which used in  skywalking oap when calling the grpc server ,  it is only defined ip and port;
         2. how to use  the uiid , should i return it in the response?  and if true ,  which position？
          
         Thanks for answer my question !

**Comment by wu-sheng** [comment]
What method name? It is a grpc service. Follow the proto we provided.

**Reply by Jiaweij** [reply]
Sorry for not asking clearly , i did not see the proto in guide and guess i need to difine it ,  where to see the proto?  find in the source code ? 
Thanks very much again !

**Reply by wu-sheng** [reply]
The link is in the first line of the doc.

https://skywalking.apache.org/docs/main/latest/en/setup/backend/dynamic-config-service/

**Comment by ranxuxin001** [comment]
hello @wu-sheng  :

I also implemented my grpc server which responses the dynamic configuration to skywalking oap. But oap log shows it calls callgroup function successfully. But my agent plugin configuration didn't change. Is my response data structure wrong?


my server responses data structure:

pb.GroupConfigurationResponse{
		GroupConfigTable: []*pb.GroupConfigItems{groupItems},
		Uuid:             s.currentGroupUUID,
	}

groupItems := &pb.GroupConfigItems{
		GroupName: "configuration-discovery.default.agentConfigurations",
	}

groupItems.Items = append(groupItems.Items, &pb.Config{
		Name:  "customerAPI-v1",
		Value: s.groupConfig,
	})

s.groupConfig := `
configurations:
  qewqw-test::myapmmysqlservice:
    agent.sample_n_per_3_secs: 1
`






data structure of official web:
groupConfigTable {
  groupName: "core.default.endpoint-name-grouping-openapi"
  items {
    name: "customerAPI-v1"
    value: "value of customerAPI-v1"
  }
  items {
    name: "productAPI-v1"
    value: "value of productAPI-v1"
  }
  items {
    name: "productAPI-v2"
    value: "value of productAPI-v2"
  }  
}

**Reply by wu-sheng** [reply, marked as answer]
The `agentConfigurations` is a **single** type configuration watcher internally (not a group type). So OAP fetches it through the `call()` RPC, not `callGroup()`.

You should return the agent configurations in your `call()` response (`ConfigurationResponse`), not in the `callGroup()` response (`GroupConfigurationResponse`). The entire YAML block should be the `value` of a single `Config` item:

```protobuf
// call() response
configTable {
  name: "configuration-discovery.default.agentConfigurations"
  value: "configurations:\n  qewqw-test::myapmmysqlservice:\n    agent.sample_n_per_3_secs: 1\n"
}
uuid: "your-uuid"
```

In your Go code, it would look like:

```go
// Implement the call() RPC, not callGroup()
response := &pb.ConfigurationResponse{
    ConfigTable: []*pb.Config{
        {
            Name:  "configuration-discovery.default.agentConfigurations",
            Value: "configurations:\n  qewqw-test::myapmmysqlservice:\n    agent.sample_n_per_3_secs: 1\n",
        },
    },
    Uuid: s.currentUUID,
}
```

The reason your current setup does not work: OAP calls both `call()` and `callGroup()` on your server periodically. But it only looks for `agentConfigurations` in the `call()` response. The `callGroup()` response is matched against group-type watchers (like `core.default.endpoint-name-grouping-openapi`), so your config there is simply ignored.

**Reply by ranxuxin001** [reply]
Thanks for your great help! I noticed skywalking official website shows group config should use core.default.endpoint-name-grouping-openapi.  And I need set agent dynamic configuration along with my service name (eg,"qewqw-test::myapmmysqlservice"). Cloud I write my callGroup()'s response like this below?
```golang
pb.GroupConfigurationResponse{
    GroupConfigTable: []*pb.GroupConfigItems{groupItems},
    Uuid:             s.currentGroupUUID,
}

groupItems := &pb.GroupConfigItems{
	GroupName: "core.default.endpoint-name-grouping-openapi", 
}

groupItems.Items = append(groupItems.Items, &pb.Config{
	Name:  "how to write",
	Value: "configurations:\n  qewqw-test::myapmmysqlservice:\n    agent.sample_n_per_3_secs: 1\n", 
})
```

**Reply by wu-sheng** [reply]
This is a misunderstanding — core.default.endpoint-name-grouping-openapi is for endpoint name grouping (OpenAPI definitions), not for agent configurations at all.
I have given a very clear reply.
